### ISSUE TITLE
Needs adapters to provide impression id associated with the request data

### ISSUE THREAD

**Issue by ShriprasadM** [prebid/prebid-server#2973]
Created separate issue as discussed in PBS-PubMatic sync up on Monday as a part of https://github.com/prebid/prebid-server/pull/2891. 

1.  As per [bidder.go - MakeRequests](https://github.com/prebid/prebid-server/blob/cc18b8a10770907988fb497cb98945f73cdc07f1/exchange/bidder.go#L135) bidder can form array of requests
https://github.com/prebid/prebid-server/blob/cc18b8a10770907988fb497cb98945f73cdc07f1/exchange/bidder.go#L135
2. So, framework gets following scenarios. But, we will not be able to match up the error (if any) corresponds to which impression id, as framework will not be able to determine the relation between `requestData` and `impressionId`
3. Scenario - Prebid Server received request with 3 impressions (Multi Impression Request) and bidder has timed out

|  Bidder | MakeRequests behaviour| Non Bid Reason |
| ------------- | ------------- | ------------- |
| Bidder A  | Return requestData array of size 3  | Imp1 - Bidder Responded, Imp2 - Bidder Timeout , Imp3 - Bidder Timeout then ideally non bid reason should be only for imp2 and imp3 and not for imp1. But, we are unable to identify the obtain array of RequestData and impId relation   |
| Bidder B  | Return requestData array of size 1  | If bidder time outs then it must be applicable for all the impressions. Because, if will single network call  |

https://github.com/prebid/prebid-server/blob/cc18b8a10770907988fb497cb98945f73cdc07f1/adapters/bidder.go#L120
```go
// RequestData packages together the fields needed to make an http.Request.
type RequestData struct {
	Method  string
	Uri     string
	Body    []byte
	Headers http.Header
        ImpID []string // new proposed field. Bidder suppose to provide array of impressions
}
```

**Comment by SyntaxNode** [comment]
Looks good, except that `ImpID string` needs to be `ImpIDs []string` since a request may contain more than one impression.

**Comment by ShriprasadM** [comment]
> Looks good, except that `ImpID string` needs to be `ImpIDs []string` since a request may contain more than one impression.

@SyntaxNode : Correct me if I am wrong, but as per this method declaration, bidder may return an `array` of ` RequestData`. My assumption here is each  `RequestData` will represent `one impression` request. 

```go
MakeRequests(request *openrtb2.BidRequest, reqInfo *ExtraRequestInfo) ([]*RequestData, []error)
```

**Comment by hhhjort** [comment]
Each `RequestData` can contain one or more impression objects. Many adapters will create one request for all the impressions. Some adapters may split the requests out into impression specific requests. Adapters may also have other ways of splitting requests, for example a bidder being set up to only be able to bid on one size per request would need to create one request for each allowed size in the bid request. I don't think we can make any assumptions on how the requests will be organized.

**Comment by ShriprasadM** [comment]
Updated  
```go       
ImpID string // new proposed field. Bidder suppose to provide
```
to 

```go
  ImpID []string // new proposed field. Bidder suppose to provide
```

**Comment by ShriprasadM** [comment]
> Looks good, except that `ImpID string` needs to be `ImpIDs []string` since a request may contain more than one impression.

Updated  
```go       
ImpID string // new proposed field. Bidder suppose to provide
```
to 

```go
  ImpID []string // new proposed field. Bidder suppose to provide
```

**Comment by SyntaxNode** [comment]
Implemented in PBS-Go v2.15.0.
